Notify networking stack that download is media related

jyavenard opened the pull request:
#### f2f0a9cc04ccbbd6ee98f86ae64ee3a3ce9ff085
<pre>
Need a short description (OOPS!).
Need the bug URL (OOPS!).

Reviewed by NOBODY (OOPS!).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
</pre>

youennf commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -8477,6 +8481,12 @@ MediaProducerMediaStateFlags HTMLMediaElement::mediaState() const
 
+#if ENABLE(MEDIA_SOURCE)
+    // We can assume that if we have active source buffers, later networking activity (such as stream or XHR requests) will be media related.
+    if (m_mediaSource && m_mediaSource->activeSourceBuffers() && m_mediaSource->activeSourceBuffers()->length())
```
What is the benefit of checking activeSourceBuffers/mediaSource over say checking that the media player is MSE and playing?

jyavenard replied:
a source buffer becomes active when an init segment has been added, this signals that more data is to come.

it would lead to faster start time with players such as YouTube than waiting to actual get data and sufficient buffer for playback to start.

youennf replied:
An alternative would be to trigger the signal when a media element is tied to a media source even though not yet playing. This would be even sooner, but maybe too aggressive with existing websites?

youennf commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10078,6 +10078,9 @@ void WebPageProxy::updatePlayingMediaDidChange(MediaProducerMediaStateFlags newS
         videoControlsManagerDidChange();
 
     m_process->updateAudibleMediaAssertions();
+    bool mediaNetworkingChanges = (oldState.contains(MediaProducerMediaState::HasNetworkingActivity)) != (newState.contains(MediaProducerMediaState::HasNetworkingActivity));
```
```suggestion
    bool mediaNetworkingChanges = oldState.contains(MediaProducerMediaState::HasNetworkingActivity) != newState.contains(MediaProducerMediaState::HasNetworkingActivity);
```

jyavenard replied:
I kept the existing "coding style" found just above :)

youennf commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1261,4 +1263,9 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CMNotificationCenter
 
+void WebProcessPool::notifyMediaNetworkingActivity(bool activity)
+{
+    [[NSNotificationCenter defaultCenter] postNotificationName:WebKitMediaNetworkingActivityNotification object:[NSNumber numberWithBool:activity]];
```
Why this mechanism over per WKWebView regular delegates?
If the goal is to send this info to network process, why not directly sending it to network process?

jyavenard replied:
because the mechanism to send current media state to the UI process was already existing.
The UI process already tracks the media state of all content processes and all pages, so re-using the infrastructure made a lot of sense.

The listener of this notification isn't the network process per say, but the radio stack on iPhone.
So having the UI process emit the notification over the network process is a non-issue

jyavenard replied:
Also the "notification-as-API" is not something we chose, it's something we have to use.

youennf commented on Tools/TestWebKitAPI/Tests/WebKit/WKPageHasMediaStreamingActivity.mm:
```diff
@@ -0,0 +1,137 @@
+    WKPageSetPageNavigationClient(page, &loaderClient.base);
+}
+
+TEST(WebKit, MSEHasMediaNetworkingActivity)
```
This test is using WK API while this is a .mm file.
If we are in a .mm file, it would make more sense to use a regular ObjC API, like `[webView loadTestPageNamed]` or `[webView stringByEvaluatingJavaScript:@"playVideo()"];`

jyavenard replied:
done

youennf commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1629,6 +1631,24 @@ void WebProcessProxy::updateAudibleMediaAssertions()
 
+void WebProcessProxy::updateMediaStreamingActivity()
+{
+    bool newMediaStreamingWebPage = WTF::anyOf(pages(), [] (auto& page) {
```
s/newMediaStreamingWebPage/hasMediaStreamingWebPage/

jyavenard replied:
I copied the style of the method above `WebProcessProxy::updateAudibleMediaAssertions.
Will modify both

youennf commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2115,6 +2116,32 @@ void WebProcessPool::updateAudibleMediaAssertions()
+    if (!m_webProcessWithMediaStreamingCounter.value()) {
+        WEBPROCESSPOOL_RELEASE_LOG(ProcessSuspension, "updateMediaStreamingActivity: The number of processes with media networking now zero. Notify network.");
+        m_mediaStreamingActivity = false;
+#if PLATFORM(COCOA)
```
I would be tempted to make the whole method #if PLATFORM(COCOA) or not at all (and make notifyMediaStreamingActivity a no-op in non cocoa ports.

jyavenard replied:
I thought about it; but it makes the change a little bit more "ugly" in my opinion as you would need to have `updateMediaStreamingActivity` present in all WebProcessPool implementation, and then `m_mediaStreamingActivity` be COCOA conditional.
Having `updateMediaStreamingActivity` virtual would have been more C++ish, but it would be a bigger change.

So ultimately, I think it's the best I can do without changing the structure too much.

jyavenard committed:
Notify networking stack that download is media related
https://bugs.webkit.org/show_bug.cgi?id=249569
rdar://103046339

Reviewed by Youenn Fablet.

Send a system notification when a networking operation is likely
going to be related to media playback.
For now the heuristic is rather primitive: if we have an active source buffer
linked to a media element we assume that all further networking operation in the
page is media related.
The notification is produced by the UI process whenever a content process
has an active source buffers, and when all have been deleted.

* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::removeSourceBuffer):
(WebCore::MediaSource::regenerateActiveSourceBuffers):
(WebCore::MediaSource::notifyElementUpdateMediaState const):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::updateMediaState):
(WebCore::HTMLMediaElement::mediaState const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/MediaProducer.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::notifyMediaNetworkingActivity):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updatePlayingMediaDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::m_webProcessWithMediaNetworkingCounter):
(WebKit::WebProcessPool::webProcessWithMediaNetworkingToken const):
(WebKit::WebProcessPool::updateMediaNetworkingActivity):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shutDown):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::updateMediaNetworkingActivity):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPageHasMediaNetworkingActivity.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKit/file-with-mse.html:

Canonical link: https://commits.webkit.org/258185@main
